A4.md

```diff
@@ -0,0 +1,254 @@
+Automatically Setting Metadata Based On Request Payload
+-------------------------------------------------------
+* Author(s): Mark D. Roth (roth@google.com)
+* Approver: a11r
+* Status: Draft
+* Implemented in: <language, ...>
+* Last updated: 2017-01-25
+* Discussion at: https://groups.google.com/d/topic/grpc-io/oOzQi1cpNfM/discussion
+
+## Abstract
+
+A mechanism for automatically setting initial metadata in a request by
+extracting information from the request payload.
+
+## Background
+
+It is often desirable to have an affinity key associated with each
+request that can be used for request routing.  Typically, this affinity
+key is based on some field from the request payload (e.g., a user name
+or a resource indicator).
+
+Because the data needed for the affinity key is in the request payload,
+it is necessary to access the deserialized request payload to generate
+the affinity key.  However, it is inefficient to deserialize the request
+payload on the server side, especially if the client is speaking to
+a proxy that is responsible for doing the routing but that would not
+otherwise need to deserialize the payload.  Thus, it makes more sense
+to extract the necessary information from the payload on the client,
+which already has the deserialized data in memory.  The affinity key can
+then be attached to the request as initial metadata, which the server
+(and/or proxy) can use for routing without having to deserialize the
+request payload.
+
+### Related Proposals:
+
+This gRFC will be followed by another one that will
+dictate how the affinity key described in this document
+will be used for affinity-based routing in [load
+balancing](https://github.com/grpc/grpc/blob/master/doc/load-balancing.md).
+
+## Proposal
+
+We will support a simple mechanism involving splitting a string on a
+given delimiter character and then recombining only a certain number of
+the resulting elements.  This will allow for common situations where the
+request names a hierarchical resource and request routing only cares
+about the highest N parts of the tree.
+
+The specification that tells us how to extract
+the headers will be configured as part of the [service
+config](https://github.com/grpc/grpc/blob/master/doc/service_config.md).
+There will also be a way to explicitly specify it via the gRPC client API.
+
+Note that for streaming RPCs, we extract headers only from the first
+request message on the stream.  However, in order to do this, we need to
+defer sending initial metadata until the client sends the first message,
+which is a change from the normal behavior.  This new behavior will only
+take effect if there is a header extraction config specified for the
+call.
+
+### Specification for Extracting Headers
```

Comment:
Could you define what happens for invalid configuration? Like, what if delimiterCharacter is two characters, or what if it is empty? Or what if the field specified payloadFieldName doesn't exist for the message? Or what if payloadFieldName is empty? What if numElementsToKeep is 0? What if headerName is an invalid header name?

I'd imagine the answer to almost all of those is going to be the same.